### ISSUE TITLE
Test failure: System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test / UseCallback_NotSecureConnection_CallbackNotCalled

### ISSUE THREAD

**Issue by karelz** [dotnet/corefx#25912]
### The handle is in the wrong state for the requested operation

Failure [20171210.03](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20171210.03/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled):
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.Http.WinHttpException : The handle is in the wrong state for the requested operation
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in E:\A\_work\1324\s\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 464
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test.<UseCallback_NotSecureConnection_CallbackNotCalled>d__8.MoveNext() in E:\A\_work\1324\s\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.ServerCertificates.cs:line 119
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at System.Threading.Tasks.RendezvousAwaitable`1.GetResult() in E:\A\_work\1324\s\corefx\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs:line 62
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext() in E:\A\_work\1324\s\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 856
```

### The operation has been canceled

Failure [20171126.02](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20171126.02/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled):
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.Http.WinHttpException : The operation has been canceled
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in E:\A\_work\686\s\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 464
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test.<UseCallback_NotSecureConnection_CallbackNotCalled>d__8.MoveNext() in E:\A\_work\686\s\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.ServerCertificates.cs:line 119
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at System.Threading.Tasks.RendezvousAwaitable`1.GetResult() in E:\A\_work\686\s\corefx\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs:line 62
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext() in E:\A\_work\686\s\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 858
```

### The handle is invalid

Failure [20171128.07](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20171128.07/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled):
```
System.Net.Http.HttpRequestException : An error occurred while sending the request.
---- System.Net.Http.WinHttpException : The handle is invalid
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in E:\A\_work\1153\s\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 464
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test.<UseCallback_NotSecureConnection_CallbackNotCalled>d__8.MoveNext() in E:\A\_work\1153\s\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.ServerCertificates.cs:line 119
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at System.Net.Http.WinHttpHandler.InternalReceiveResponseHeadersAsync(WinHttpRequestState state) in E:\A\_work\1153\s\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 1365
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext() in E:\A\_work\1153\s\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 863
```

## History of failures

Day | Build | OS | Error message
-- | -- | -- | --
5/19 | 20170519.03 | Win10 | The operation has been canceled
8/22 | 20170822.01 | Win7 | The operation has been canceled
8/31 | 20170831.01 | Win10 | The operation has been canceled
9/26 | 20170926.06 | Win7 | The handle is in the wrong state for the requested operation
10/3 | 20171003.01 | Win7 | The handle is in the wrong state for the requested operation
10/4 | 20171004.01 | Win7 | The handle is in the wrong state for the requested operation
10/22 | 20171022.01 | Win8.1 | The operation has been canceled
11/3 | 20171103.07 | Win7 | The handle is in the wrong state for the requested operation
11/3 | 20171103.11 | Win7 | The handle is in the wrong state for the requested operation
11/4 | 20171104.04 | Win7 | The handle is in the wrong state for the requested operation
11/6 | 20171106.02 | Win8.1 | The operation has been canceled
11/9 | 20171109.05 | Win7 | The handle is in the wrong state for the requested operation
11/19 | 20171119.02 | Win8.1 | The operation has been canceled
11/26 | 20171126.02 | Win8.1 | The operation has been canceled
11/28 | 20171128.07 | Win10 | The handle is invalid
12/10 | 20171210.03 | Win7 | The handle is in the wrong state for the requested operation
1/2 | 20180102.02 | Win7 | The handle is in the wrong state for the requested operation<br/>Note: Only in [catastrophic failure logs](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180102.02/workItem/System.Net.Http.Functional.Tests/wilogs)
1/11 | 20180111.01 | Win7 | The handle is in the wrong state for the requested operation - [link](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180111.01/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled)
1/11 | 20180111.02 | Win7 | The handle is in the wrong state for the requested operation - [link](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180111.02/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled)
1/15 | 20180115.04 | Win7 | The handle is invalid - [link](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180115.04/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled)
1/22 | 20180122.02 | Win7 | The handle is in the wrong state for the requested operation<br/>Note: Only in [catastrophic failure logs](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180122.02/workItem/System.Net.Http.Functional.Tests/wilogs)
2/6 | 20180206.05 | Win7 | The handle is invalid - [link](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20180206.05/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_ServerCertificates_Test~2FUseCallback_NotSecureConnection_CallbackNotCalled)

Similar exceptions / errors as in #25896 and #25902

**Comment by davidsh** [comment]
The root cause of most of the "handle is invalid" errors is:
https://github.com/dotnet/corefx/issues/9228

I have some ideas of how to fix this in WinHttpHandler/HttpClientHandler but need to prioritize this bug higher so that it gets fixed.

**Comment by karelz** [comment]
@davidsh if the root cause #9228 is fixed, are you confident this test would pass? Or could it just transform the exception into another one (e.g. the "The operation has been canceled")?

Does your root-cause statement apply just to "The handle is invalid" (1 hit in above history), or also to "The handle is in the wrong state for the requested operation" (8 hits)?

**Comment by davidsh** [comment]
>@davidsh if the root cause #9228 is fixed, are you confident this test would pass when it hits the "handle is invalid" path? Or could it just transform the exception into another one (e.g. the "The operation has been canceled")?

It would transform "handle is invalid" to a normal "cancellation" / "timeout" exception.

>Does your root-cause statement apply just to "The handle is invalid" (1 hit in above history), or also to "The handle is in the wrong state for the requested operation" (8 hits)?

Not sure.  Those are 2 different WinHTTP errors.  They might be related to same root cause or not.  Need to investigate that.

**Comment by karelz** [comment]
Thanks! That means we should keep counting the "handle is invalid" hits as part of the test failures.
Only the way the error is surfaced will be addressed by #9228. We still need to make the test robust against such situations (or fix the product even more).

**Comment by rmkerr** [comment]
Closing as this failure has not occurred since March 8th.
